yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +100,179 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
     if (credentials != null) credentials.getSecretKey(new Text(key)) else null
   }
 
+  private[spark] override def scheduleLoginFromKeytab(): Unit = {
+    sparkConf.getOption("spark.yarn.principal").foreach { principal =>
+      val keytab = sparkConf.get("spark.yarn.keytab")
+      val delegationTokenRenewerRunnable =
+        new Runnable {
+          override def run(): Unit = {
+            renewCredentials(principal, keytab)
+            delegationTokenRenewer.schedule(
+              this, (0.75 * (getLatestValidity - System.currentTimeMillis())).toLong,
+              TimeUnit.MILLISECONDS)
+          }
+        }
+      val timeToRenewal = (0.75 * (getLatestValidity - System.currentTimeMillis())).toLong
+      delegationTokenRenewer.schedule(
+        delegationTokenRenewerRunnable, timeToRenewal, TimeUnit.MILLISECONDS)
+    }
+  }
+
+  private def renewCredentials(principal: String, keytab: String): Unit = {
```

Comment:
Could you add a comment somewhere in the code about how the credentials update work? That will make it easier for someone trying to understand this code later.